File: src/updater.py
import os
import sys
import logging
import requests
import shutil
from pathlib import Path
from typing import Optional
from dataclasses import dataclass

@dataclass
class Release:
    version: str
    download_url: str
    
class Updater:
    def __init__(self, app_dir: Path, repo: str):
        self.app_dir = app_dir
        self.api_url = f"https://api.github.com/repos/{repo}/releases/latest"
        self.temp_dir = app_dir / 'temp'
        self.temp_dir.mkdir(exist_ok=True)

    def check_update(self) -> Optional[Release]:
        try:
            resp = requests.get(self.api_url)
            if resp.status_code == 200:
                data = resp.json()
                return Release(
                    version=data['tag_name'],
                    download_url=data['assets'][0]['browser_download_url']
                )
        except Exception:
            return None

    def update(self, release: Release) -> bool:
        try:
            exe = self._download(release.download_url)
            if exe:
                shutil.move(exe, self.app_dir / 'WebDesk.exe')
                return True
        except Exception:
            return False

    def _download(self, url: str) -> Optional[Path]:
        temp_file = self.temp_dir / 'update.exe'
        with requests.get(url, stream=True) as r:
            with open(temp_file, 'wb') as f:
                shutil.copyfileobj(r.raw, f)
        return temp_file

class PortableUpdater:
    def __init__(self):
        self.exe_path = Path(sys.executable)
        self.app_dir = self.exe_path.parent
        self.temp_dir = self.app_dir / 'temp'
        self.config_backup = None

    def update(self, release_info: dict) -> bool:
        try:
            # Backup config
            self._backup_config()
            
            # Download and replace executable
            new_exe = self._download_update(release_info['download_url'])
            if new_exe:
                # Replace executable
                shutil.move(new_exe, self.exe_path)
                
                # Restore config
                self._restore_config()
                
                # Restart application
                self._restart()
                return True
            return False
        except Exception:
            self._restore_config()
            return False

    def _backup_config(self):
        """Backup user configuration"""
        config_path = self.app_dir / 'config.json'
        if config_path.exists():
            self.config_backup = config_path.read_bytes()

    def _restore_config(self):
        """Restore user configuration"""
        if self.config_backup:
            config_path = self.app_dir / 'config.json'
            config_path.write_bytes(self.config_backup)

    def _download_update(self, url: str) -> Optional[Path]:
        """Download update preserving user data"""
        self.temp_dir.mkdir(exist_ok=True)
        temp_exe = self.temp_dir / 'WebDesk.exe.new'
        
        with requests.get(url, stream=True) as r:
            with open(temp_exe, 'wb') as f:
                shutil.copyfileobj(r.raw, f)
        return temp_exe

    def _restart(self):
        """Restart application"""
        os.execl(str(self.exe_path), str(self.exe_path), *sys.argv[1:])

File: src/Updater.cs
using System.Diagnostics;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text.Json;

namespace WebDesk;

internal sealed record UpdateInfo(string Version, string DownloadUrl, string Changelog);

/// <summary>
/// GitHub release self-updater. Unlike the Python original (which tried to
/// overwrite a running exe and never restarted cleanly), this renames the
/// running binary, swaps in the new one, and relaunches.
/// </summary>
internal static class Updater
{
    private static readonly HttpClient Http = CreateClient();

    private static HttpClient CreateClient()
    {
        var c = new HttpClient { Timeout = TimeSpan.FromSeconds(15) };
        c.DefaultRequestHeaders.UserAgent.Add(new ProductInfoHeaderValue("WebDesk", "1.0"));
        c.DefaultRequestHeaders.Accept.Add(
            new MediaTypeWithQualityHeaderValue("application/vnd.github+json"));
        return c;
    }

    public static async Task<UpdateInfo?> CheckAsync(string repo, string currentVersion)
    {
        try
        {
            var url = $"https://api.github.com/repos/{repo}/releases/latest";
            using var resp = await Http.GetAsync(url);
            if (!resp.IsSuccessStatusCode) return null;

            using var doc = JsonDocument.Parse(await resp.Content.ReadAsStringAsync());
            var root = doc.RootElement;

            var tag = root.GetProperty("tag_name").GetString() ?? "";
            var body = root.TryGetProperty("body", out var b) ? b.GetString() ?? "" : "";

            string? asset = null;
            if (root.TryGetProperty("assets", out var assets) && assets.GetArrayLength() > 0)
                asset = assets[0].GetProperty("browser_download_url").GetString();
            if (asset is null) return null;

            return IsNewer(tag, currentVersion) ? new UpdateInfo(tag, asset, body) : null;
        }
        catch
        {
            return null;
        }
    }

    private static bool IsNewer(string tag, string current)
    {
        static Version Parse(string s)
        {
            s = s.TrimStart('v', 'V').Trim();
            return Version.TryParse(s, out var v) ? v : new Version(0, 0);
        }

        return Parse(tag) > Parse(current);
    }

    public static async Task ApplyAsync(UpdateInfo update)
    {
        var exe = Environment.ProcessPath
                  ?? throw new InvalidOperationException("Cannot locate the current executable.");
        var newExe = exe + ".new";
        var oldExe = exe + ".old";

        using (var resp = await Http.GetAsync(update.DownloadUrl, HttpCompletionOption.ResponseHeadersRead))
        {
            resp.EnsureSuccessStatusCode();
            await using var fs = File.Create(newExe);
            await resp.Content.CopyToAsync(fs);
        }

        // Windows won't overwrite a running exe, but it will rename it.
        if (File.Exists(oldExe)) File.Delete(oldExe);
        File.Move(exe, oldExe);
        File.Move(newExe, exe);

        Process.Start(new ProcessStartInfo(exe) { UseShellExecute = true });
    }

    public static void CleanupOldVersion()
    {
        try
        {
            if (Environment.ProcessPath is not { } exe) return;
            var oldExe = exe + ".old";
            if (File.Exists(oldExe)) File.Delete(oldExe);
        }
        catch
        {
            // The previous process may still hold a lock; it'll clear next run.
        }
    }
}
